pyccel/ast/cudaext.py

```diff
@@ -0,0 +1,309 @@
+class CudaArray(CudaNewArray):
+    """
+    Represents a call to  cuda.array for code generation.
+
+    arg : list, tuple, PythonList
+
+    """
+    __slots__ = ('_arg','_dtype','_precision','_shape','_rank','_order','_memory_location')
+    _attribute_nodes = ('_arg',)
+    name = 'array'
+
+    def __init__(self, arg, dtype=None, order='C', memory_location='managed'):
+
+        if not isinstance(arg, (PythonTuple, PythonList, Variable)):
+            raise TypeError('Unknown type of  %s.' % type(arg))
+
+        is_homogeneous_tuple = isinstance(arg, (PythonTuple, PythonList, HomogeneousTupleVariable)) and arg.is_homogeneous
+        is_array = isinstance(arg, Variable) and arg.is_ndarray
+
+        # TODO: treat inhomogenous lists and tuples when they have mixed ordering
+        if not (is_homogeneous_tuple or is_array):
+            raise TypeError('we only accept homogeneous arguments')
+
+        # Verify dtype and get precision
+        if dtype is None:
+            dtype = arg.dtype
+            prec = get_final_precision(arg)
+        else:
+            dtype, prec = process_dtype(dtype)
+        # ... Determine ordering
+        order = str(order).strip("\'")
+
+        shape = process_shape(False, arg.shape)
+        rank  = len(shape)
+
+        if rank < 2:
+            order = None
+        else:
+            # ... Determine ordering
+            order = str(order).strip("\'")
+
+            if order not in ('K', 'A', 'C', 'F'):
+                raise ValueError(f"Cannot recognize '{order}' order")
+
+            # TODO [YG, 18.02.2020]: set correct order based on input array
+            if order in ('K', 'A'):
+                order = 'C'
+            # ...
+        #Verify memory location
+        if memory_location not in ('host', 'device', 'managed'):
+            raise ValueError("memory_location must be 'host', 'device' or 'managed'")
+        self._arg   = arg
+        self._shape = shape
+        self._rank  = rank
+        self._dtype = dtype
+        self._order = order
+        self._precision = prec
+        self._memory_location = memory_location
+        super().__init__()
+
+    def __str__(self):
+        return str(self.arg)
+
+    @property
+    def arg(self):
+        return self._arg
+    @property
+    def memory_location(self):
+        return self._memory_location
+
+class CudaSynchronize(PyccelInternalFunction):
+    "Represents a call to  Cuda.deviceSynchronize for code generation."
+
+    __slots__ = ('_dtype','_precision','_shape','_rank','_order')
+    _attribute_nodes = ()
+    def __init__(self):
+        #...
+        self._shape     = None
+        self._rank      = 0
+        self._dtype     = NativeInteger()
+        self._precision = None
+        self._order     = None
+        super().__init__()
+
+class CudaInternalVar(PyccelAstNode):
+    """
+    Represents a General Class For Cuda internal Variables Used To locate Thread In the GPU architecture"
```

Comment:
I don't understand the purpose of this class. Is `CudaInternalVar(LiteralInteger(2))` equivalent to `Variable('int','X', is_temp=True)`? Or is it not printed? If it isn't printed then is a class needed for this? It only seems to store `0`/`1`/`2` and doesn't do anything more complicated